File: src/com/jda/algorithm_programs/SearchAndSort.java
/*******************************************************************
 * Purpose:
 *
 * @author Mohan Pednekar
 * @version 1.0
 * @since Jul 11, 2018
 *******************************************************************/
package com.jda.algorithm_programs;

import com.jda.utility.Util.Stopwatch;
import com.jda.utility.Utility;
import com.jda.utility.Utility.Sorting;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.HashMap;

public
class SearchAndSort<T1> implements Comparator<T1> {

  public static
  void main(String[] args) {
    Utility utility = new Utility();
    SearchAndSort programs = new SearchAndSort<>();
    utility.printLine("What do you want to enter?");
    utility.printLine("s : Strings \t i : Integers");
    char ch = utility.readString().toLowerCase().charAt(0);
    utility.printLine("How many of them?");
    int n = utility.readInteger();

    switch (ch) {
      case 's':
        ArrayList<String> strings = new ArrayList<>();
        utility.printLine("Enter " + n + " strings");

        for (int i = 0; i < n; i++) {
          strings.add(utility.readString());
        }
        utility.printLine("What do you want to search for?");
        String stringToSearch = utility.readString();
        programs.searchAndSort(strings, stringToSearch);
        break;
      case 'i':
        ArrayList<Integer> integers = new ArrayList<>();
        utility.printLine("Enter " + n + " integers");
        for (int i = 0; i < n; i++) {
          integers.add(utility.readInteger());
        }
        utility.printLine("What do you want to search for?");
        Integer integerToSearch = utility.readInteger();
        programs.searchAndSort(integers, integerToSearch);
        break;
      default:
    }
  }

  /**
   * @param items
   */
  private
  <T> void searchAndSort(ArrayList<T> items, T itemToSearch) {

    Watcher.BINARY.start();
    boolean found = binarySearch(items, itemToSearch);
    Watcher.BINARY.stop();
    System.out.println("found = " + found);

    Watcher.INSERTION.start();
    ArrayList<T> insertionSorted = insertionSort(items);
    Watcher.INSERTION.stop();
    System.out.println(insertionSorted);

    Watcher.BUBBLE.start();
    ArrayList<T> bubbleSorted = Sorting.bubbleSort(items);
    Watcher.BUBBLE.stop();
    System.out.println(bubbleSorted);

    HashMap<Watcher, Long> watchersMap = new HashMap<>();

    for (Watcher watcher : Watcher.values()) {
      System.out.println(watcher.resultString());
      watchersMap.put(watcher, watcher.elapsedTimeNanos());
    }
  }

  /**
   * @param items
   * @param itemToSearch
   * @return
   */
  private
  <T> boolean binarySearch(ArrayList<T> items, T itemToSearch) {
    BST tree = new BST(items.get(0));
    items.forEach(tree::insert);
    return tree.has(itemToSearch);
  }

  /**
   * @param items
   * @return
   */
  private
  <T> ArrayList<T> insertionSort(ArrayList<T> items) {
    int n = items.size();
    for (int i = 0; i < n; i++) {
      for (int j = i; j > 0; j--) {
        T1 i1 = (T1) items.get(j - 1);
        T1 i2 = (T1) items.get(j);
        if (compare(i1, i2) > 0) {
          Collections.swap(items, j - 1, j);
        }
      }
    }
    return items;
  }  /*
   * (non-Javadoc)
   * @see java.util.Comparator#compare(java.lang.Object, java.lang.Object)
   */

  @Override
  public
  int compare(T1 t, T1 t2) {
    if (t.getClass().toString().endsWith("Integer")) {
      return ((Integer) t).compareTo((Integer) t2);
    }
    if (t.getClass().toString().endsWith("String")) {
      return ((String) t).compareToIgnoreCase((String) t2);
    }
    return 0;
  }

  enum Watcher {
    BINARY("Binary search"), INSERTION("Insertion sort"), BUBBLE("Bubble sort");
    Stopwatch stopwatch = new Stopwatch();
    String watcherName;

    Watcher(String str) {
      watcherName = str;
    }

    public
    void start() {
      stopwatch.start();
    }

    public
    void stop() {
      stopwatch.stop();
    }

    public
    long elapsedTimeNanos() {
      return stopwatch.elapsedTimeNanos();
    }

    public
    String resultString() {
      return watcherName + " took " + elapsedTimeSeconds() + " seconds";
    }

    public
    double elapsedTimeSeconds() {
      return stopwatch.elapsedTimeSeconds();
    }
  }

  public
  class BST<T> implements Comparator<T> {

    T value;
    BST left;
    BST right;

    BST(T value) {
      this.value = value;
    }

    /*
     * (non-Javadoc)
     * @see java.util.Comparator#compare(java.lang.Object, java.lang.Object)
     */
    @Override
    public
    int compare(T o1, T o2) {
      if (o1.getClass().toString().endsWith("Integer")) {
        return ((Integer) o1).compareTo((Integer) o2);
      }
      if (o1.getClass().toString().endsWith("String")) {
        return ((String) o1).compareToIgnoreCase((String) o2);
      }
      return 0;
    }

    public
    boolean has(T value) {
      if (compare(value, this.value) == 0) {
        return true;
      }
      if (compare(value, this.value) < 0 && null != left) {
        return left.has(value);
      }
      if (compare(value, this.value) > 0 && null != right) {
        return right.has(value);
      }
      return false;
    }

    public
    void insert(T value) {
      if (compare(value, this.value) <= 0) {
        if (null == left) {
          left = new BST(value);
        } else {
          left.insert(value);
        }
      } else if (null == right) {
        right = new BST(value);
      } else {
        right.insert(value);
      }
    }
  }


}

File: src/com/jda/algorithm_programs/SearchAndSort.kt
/*******************************************************************
 * Purpose:
 *
 * @author Mohan Pednekar
 * @version 1.0
 * @since Jul 11, 2018
 */
package com.jda.algorithm_programs

import com.jda.utility.Util.Stopwatch
import com.jda.utility.Utility
import com.jda.utility.Utility.Sorting
import java.util.*
import java.util.function.Consumer

class SearchAndSort {

    /**
     * @param items
     */
    private fun <T : Comparable<T>> searchAndSort(items: ArrayList<T>, itemToSearch: T) {

        Watcher.BINARY.start()
        val found = binarySearch(items, itemToSearch)
        Watcher.BINARY.stop()
        println("found = $found")

        Watcher.INSERTION.start()
        val insertionSorted = Sorting.insertionSort(items)
        Watcher.INSERTION.stop()
        println(insertionSorted)

        Watcher.BUBBLE.start()
        val bubbleSorted = Sorting.bubbleSort(items)
        Watcher.BUBBLE.stop()
        println(bubbleSorted)

        val watchersMap = HashMap<Watcher, Long>()

        for (watcher in Watcher.values()) {
            println(watcher.resultString())
            watchersMap[watcher] = watcher.elapsedTimeNanos()
        }
    }

    /**
     * @param items
     * @param itemToSearch
     * @return
     */
    fun <T : Comparable<T>> binarySearch(items: ArrayList<T>, itemToSearch: T): Boolean {
        val tree = BST(items[0])
        items.forEach(Consumer<T> { tree.insert(it) })
        return tree.has(itemToSearch)
    }

    internal enum class Watcher(var watcherName: String) {
        BINARY("Binary search"), INSERTION("Insertion sort"), BUBBLE("Bubble sort");

        var stopwatch = Stopwatch()

        fun start() {
            stopwatch.start()
        }

        fun stop() {
            stopwatch.stop()
        }

        fun elapsedTimeNanos(): Long {
            return stopwatch.elapsedTimeNanos()
        }

        fun resultString(): String {
            return watcherName + " took " + elapsedTimeSeconds() + " seconds"
        }

        fun elapsedTimeSeconds(): Double {
            return stopwatch.elapsedTimeSeconds()
        }
    }

    class BST<T : Comparable<T>> internal constructor(private var value: T) {
        private var left: BST<T>? = null
        private var right: BST<T>? = null

        /*
     * (non-Javadoc)
     * @see java.util.Comparator#compare(java.lang.Object, java.lang.Object)
     */
        //    @Override
        //    public
        //    int compare(T o1, T o2) {
        //      if (o1.getClass().toString().endsWith("Integer")) {
        //        return ((Integer) o1).compareTo((Integer) o2);
        //      }
        //      if (o1.getClass().toString().endsWith("String")) {
        //        return ((String) o1).compareToIgnoreCase((String) o2);
        //      }
        //      return 0;
        //    }
        fun has(value: T): Boolean {
            if (value.compareTo(this.value) == 0) {
                return true
            }
            if (value.compareTo(this.value) < 0 && null != left) {
                return left!!.has(value)
            }
            return if (value.compareTo(this.value) > 0 && null != right) {
                right!!.has(value)
            } else false
        }

        fun insert(value: T) {
            if (value.compareTo(this.value) <= 0) {
                if (null == left) {
                    left = BST(value)
                } else {
                    left!!.insert(value)
                }
            } else if (null == right) {
                right = BST(value)
            } else {
                right!!.insert(value)
            }
        }
    }

    companion object {

        @JvmStatic
        fun main(args: Array<String>) {
            val reader = Utility().Reader()
            val programs = SearchAndSort()
            Utility.printLine("What do you want to enter?")
            Utility.printLine("s : Strings \t i : Integers")
            val ch = reader.readString().toLowerCase().get(0)
            Utility.printLine("How many of them?")
            val n = reader.readInteger()

            when (ch) {
                's' -> {
                    val strings = ArrayList<String>()
                    Utility.printLine("Enter $n strings")

                    for (i in 0 until n) {
                        strings.add(reader.readString())
                    }
                    Utility.printLine("What do you want to search for?")
                    val stringToSearch = reader.readString()
                    programs.searchAndSort(strings, stringToSearch)
                }
                'i' -> {
                    val integers = ArrayList<Int>()
                    Utility.printLine("Enter $n integers")
                    for (i in 0 until n) {
                        integers.add(reader.readInteger())
                    }
                    Utility.printLine("What do you want to search for?")
                    val integerToSearch = reader.readInteger()
                    programs.searchAndSort(integers, integerToSearch)
                }
            }
        }
    }
}
